libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayProgressObserver.kt

```diff
@@ -38,20 +44,58 @@ class ReplayProgressObserver(
      */
     override fun onRouteProgressChanged(routeProgress: RouteProgress) {
-        val routeProgressRouteLeg = routeProgress.currentLegProgress?.routeLeg
-        if (routeProgressRouteLeg != currentRouteLeg) {
+        val currentLegProgress = routeProgress.currentLegProgress
+        val routeProgressRouteLeg = currentLegProgress?.routeLeg
+        if (routeProgressRouteLeg != currentRouteLeg && currentLegProgress != null) {
             this.currentRouteLeg = routeProgressRouteLeg
-            onRouteLegChanged(routeProgressRouteLeg)
+            onRouteLegChanged(routeProgressRouteLeg, currentLegProgress.distanceTraveled)
         }
     }
 
-    private fun onRouteLegChanged(routeProgressRouteLeg: RouteLeg?) {
-        if (routeProgressRouteLeg != null) {
-            val replayEvents = replayRouteMapper.mapRouteLegGeometry(routeProgressRouteLeg)
-            if (replayEvents.isNotEmpty()) {
-                mapboxReplayer.pushEvents(replayEvents)
-                mapboxReplayer.seekTo(replayEvents.first())
+    private fun onRouteLegChanged(routeProgressRouteLeg: RouteLeg?, distanceTraveled: Float?) {
+        if (routeProgressRouteLeg == null) return
+        val replayEvents = replayRouteMapper.mapRouteLegGeometry(routeProgressRouteLeg)
+        if (replayEvents.isEmpty()) return
+        val shouldClipRoute = distanceTraveled != null && distanceTraveled > CLIP_THRESHOLD_METERS
+        val seekToIndex = if (shouldClipRoute) indexAlong(replayEvents, distanceTraveled!!) else 0
+        mapboxReplayer.pushEvents(replayEvents)
+        mapboxReplayer.seekTo(replayEvents[seekToIndex])
+    }
+
+    /**
+     * Measures the distance between the [events] and returns the index of an event that is greater
+     * than the [distanceTraveled]. If the distanceTraveled is greater than the event distance, the
+     * last index is returned. If the event list is empty -1 is returned.
+     *
+     * events: 0---1-------2------3--4-----5------
+     *         ==distanceTraveled====>||
+     * returns index: 5
+     */
+    private fun indexAlong(events: List<ReplayEventBase>, distanceTraveled: Float): Int {
+        var currentDistance = 0.0
+        var index = -1
+        var lastPoint: Point? = null
+        for (item in events) {
+            if (item is ReplayEventUpdateLocation) {
+                val currentPoint = Point.fromLngLat(item.location.lon, item.location.lat)
+                lastPoint?.let {
+                    currentDistance += TurfMeasurement.distance(
+                        it,
+                        currentPoint,
+                        TurfConstants.UNIT_METERS
+                    )
+                }
+                lastPoint = currentPoint
+            }
+            index++
+            if (currentDistance >= distanceTraveled) {
+                return index
             }
         }
+        return index
+    }
+
+    private companion object {
+        private const val CLIP_THRESHOLD_METERS = 10.0
```

Comment:
Why is the clip needed? Looks like the algo would succeed even without it, wouldn't it?

Reply:
I'll remove this. Yeah it works without it. Also 10 meters is large enough to go backwards when simulating traffic.

I noticed in logs that `distanceTraveled` can have small floating point  "onRouteLegChanged 1.0E-5" so made a initial case. It didn't end up impacting any tests though, so think you're right to remove.

But still need to handle the null case, `distanceTraveled ?: 0.0f` works.

Reply:
actually null is not needed either :)